### PR TITLE
Implement evaluation of an algorithm and a trained model

### PR DESCRIPTION
**Pull request by blake-wulfe-tri**
This commit glues together the code to run eval.

- - -
This change is [Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)

### CODE REVIEW

**Review by blake-wulfe-tri — COMMENTED** [review]
+@hongkai-dai for review please!

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-:-NPPtCOh90eMctw0nRcD:b-jjzuhe)* status: 0 of 17 files reviewed, all discussions resolved (waiting on @hongkai-dai)

**Review by blake-wulfe-tri — COMMENTED** [review]
I'm still not confident that this all works yet. Need to implement some visualization / debugging tools first.

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-:-NPPtF3X9ldgqGxiUnJU:bdxge35)* status: 0 of 17 files reviewed, all discussions resolved (waiting on @hongkai-dai)

**Review by hongkai-dai — CHANGES_REQUESTED** [review]
> Reviewed 15 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)* status: 15 of 17 files reviewed, 2 unresolved discussions (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/algorithms/action_samplers.py` line 28 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPQbjy9AShTMLIL1CHd:-NPQbjy9AShTMLIL1CHe:bxvmfyb) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/algorithms/action_samplers.py#L28)):*
> ```Python
>             height_lb: Lower bound of the height to sample.
>             height_ub: Upper bound of the height to sample.
>             rolling_distance: Distance to roll.
> ```

I would think that we will also sample the rolling distance?
___
*[`mani_skill2/dough_model_learning/generate_dough_rolling_data.py` line 34 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPQd2qI9CgJVbZAZlfp:-NPQd2qI9CgJVbZAZlfq:bqacven) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/dough_model_learning/generate_dough_rolling_data.py#L34)):*
> ```Python
> 
> 
> class CircularConeHeightMapGenerator:
> ```

I am not familiar with this, why using a class with `__call__` function instead of the original function?

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 1 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)* status: 16 of 17 files reviewed, 2 unresolved discussions (waiting on @blake-wulfe-tri)

**Review by hongkai-dai — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)* status: 16 of 17 files reviewed, 3 unresolved discussions (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dough_model_learning/eval.py` line 56 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPT3jro7GNXeGDRB8FY:-NPT3jro7GNXeGDRB8FZ:bwoxs94) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/dough_model_learning/eval.py#L56)):*
> ```Python
>     # algorithm = hydra.utils.instantiate(cfg.eval.algorithm)
>     algorithm = get_algo(env, cfg)
>     rollouts = generate_rollouts(
> ```

Should we add an `eval` mode to `generate_rollouts`, that we disable the gradient computation (using torch.no_grad())?

**Comment by hongkai-dai** [comment]
I don't quite understand the CI failure, it complains about not finding `mani_skill2.algorithms.action_samplers`, but I do see this file. On the other hand, it doesn't complain about the other files in the same folder, like `mani_skill2.algorithms.gradient_shooting`.

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)* status: 16 of 17 files reviewed, 3 unresolved discussions (waiting on @hongkai-dai)

___
*[`mani_skill2/algorithms/action_samplers.py` line 28 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPQbjy9AShTMLIL1CHd:-NPTJW04AJbPu0fpxbPb:b-daajby) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/algorithms/action_samplers.py#L28)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I would think that we will also sample the rolling distance?
</blockquote></details>

hmm yeah I had a todo above, but I think it's probably better to just have the bounds for everything but set them to the specific value for fixed ones for now
___
*[`mani_skill2/dough_model_learning/eval.py` line 56 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPT3jro7GNXeGDRB8FY:-NPTJoxX-Kkkr_3NL54q:b-rtj4qp) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/dough_model_learning/eval.py#L56)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Should we add an `eval` mode to `generate_rollouts`, that we disable the gradient computation (using torch.no_grad())?
</blockquote></details>

yeah good point, I'll add that
___
*[`mani_skill2/dough_model_learning/generate_dough_rolling_data.py` line 34 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPQd2qI9CgJVbZAZlfp:-NPTK1xy0mdnCwehsDJc:b7ir2a1) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/dough_model_learning/generate_dough_rolling_data.py#L34)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I am not familiar with this, why using a class with `__call__` function instead of the original function?
</blockquote></details>

I did this just to make `hydra.instantiate` easier to use

**Review by blake-wulfe-tri — COMMENTED** [review]
hmm ok will look into this later

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-:-NPTK6aS8GGfPa9K1Ctu:bqswlkg)* status: 16 of 17 files reviewed, 3 unresolved discussions (waiting on @hongkai-dai)

**Review by hongkai-dai — COMMENTED** [review]
> Reviewed 1 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @blake-wulfe-tri)

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @hongkai-dai)

___
*[`mani_skill2/dough_model_learning/eval.py` line 56 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPT3jro7GNXeGDRB8FY:-NPYlkXK2O9R8m7XESNk:ba2mleh) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/dough_model_learning/eval.py#L56)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

yeah good point, I'll add that
</blockquote></details>

Actually I'm not sure if we should do that. e.g., if one of the planning methods use gradients.

**Review by hongkai-dai — COMMENTED** [review]
I found that if I try to set `export PYTHONPATH=PATH_TO_CHECKED_OUT_MANISKILL2` then the CI can succeed and find the right package for import.

This surprised me because I think python will automatically look into the current directory to find the package, and I am in that `PATH_TO_CHECKED_OUT_MANISKILL2` directory when running `pytest`.

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-:-NPZIJ_o4GEipsWIbxwH:be3dowj)* status: 16 of 17 files reviewed, all discussions resolved (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dough_model_learning/eval.py` line 56 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-NPT3jro7GNXeGDRB8FY:-NPZIDiB5z-qyfiFzx2j:bv0tszw) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/b782b26a500e85b94c37d0402081a24598842dc3/mani_skill2/dough_model_learning/eval.py#L56)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

Actually I'm not sure if we should do that. e.g., if one of the planning methods use gradients.
</blockquote></details>

Good point. I mark myself satisfied.

**Review by hongkai-dai — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/22#-:-NPZIhFy32TKSTdzBU0l:bnfp4nl)* status: 16 of 17 files reviewed, all discussions resolved (waiting on @blake-wulfe-tri)
